include/boost/geometry/iterators/detail/segment_iterator/range_segment_iterator.hpp

```diff
@@ -100,23 +100,40 @@ class range_segment_iterator
             Reference
         >
 {
+private:
+    static bool has_less_than_two_elements(Range const& r)
+    {
+        iterator_type first = range_iterator_begin<Range>::apply(r);
+        iterator_type last = range_iterator_end<Range>::apply(r);
+
+        return first == last || ++first == last;
```

Comment:
If that is the case. Then I think boost::size(r) < 2u will also do at the caller side?
We don't need this function then at all. That looks better to me.
All (or at least most of) our functionality is based on random access ranges, accessing them with segment-id's and so on, I don't think we have to provide complexity to support non random access ranges here.